TRD: Tracker: Add padrow crossing information

f3sch opened the pull request:
This patch adds to the GPUTRDTrack one new members:
     1. mIsPadrowCrossing:
        unsigned char -> bits 0 to 5 indicate if in a layer the track
        crossed a padrow and bit 6 indicate whether a neighborign was found.
        Rational: Charge information is lost in these tracklets and
        one should possibly even discard those Tracks for gain
        calibration and PID.

Additionally, this was added to TrackQC.

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

f3sch committed:
TRD: Tracker: Add padrow crossing information

This patch adds to the GPUTRDTrack one new members:
     1. mIsPadrowCrossing:
        unsigned char -> bits 0 to 5 indicate if in a layer the track
        crossed a padrow.
        Rational: Charge information is lost in these tracklets and
        one should possibly even discard those Tracks for gain
        calibration and PID.

Additionally, this was added to TrackQC.

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

shahor02 commented on GPU/GPUTracking/DataTypes/GPUTRDTrack.h:
```diff
@@ -106,13 +107,15 @@ class GPUTRDTrack_t : public T
+  unsigned char mIsPadrowCrossing;  // bits 0 to 5 indicate wheter track was in between two padrows
   unsigned int mRefGlobalTrackId;   // raw GlobalTrackID of the seeding track (either ITS-TPC or TPC)
   int mAttachedTracklets[kNLayers]; // indices of the tracklets attached to this track; -1 means no tracklet in that layer
   short mCollisionId;               // the collision ID of the tracklets attached to this track; is used to retrieve the BC information for this track after the tracking is done
```
Could you move mCollisionID before mAttached..., to not lose 2 bytes on the alignment?

f3sch replied:
Sure, did not know that alignment was a concern here. Sorry.
But from my very limited understanding of the topic, the situation is like this:

```
// size = 36, alignment = 4
// before:  +----+----+-*6-+----+
// members: |aaaa|bbbb|c**c|dde.|
// . represents a byte of padding
// a=mChi2, b=mRefGlobalTrackId, c=mAttachedTracklets, d=mCollisionId, e=mFlags
```

```
// size = 40, alignment = 4
// mine:    +----+----+----+-*6-+----+
// members: |aaaa|p...|bbbb|c**c|dde.|
// p=mIsPadrowCrossing
```

```
// size = 44, alignment = 4
// yours:   +----+----+----+----+-*6-+----+
// members: |aaaa|p...|bbbb|dd..|c**c|e....|
```

Maybe better is:
```
// size = 36, alignment = 4
// mine:    +----+----+-*6-+----+
// members: |aaaa|bbbb|c**c|ddep|
```
e.g. moving mIsPadrowCrossing after mFlags.

At least that what I get from testing in an interactive root session and defining
struct foo {...} and playing with alignof() and sizeof() on x86, not sure if this is valid since this is run on a GPU?

martenole replied:
I think this was just a typo. Putting `mIsPadrowCrossing` after `mFlags` is fine

martenole replied:
Just not to put a 1 byte member between 2 4 byte members

f3sch replied:
Done.

shahor02 commented on GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:
```diff
@@ -657,6 +657,22 @@ GPUd() bool GPUTRDTracker_t<TRDTRK, PROP>::FollowProlongation(PROP* prop, TRDTRK
+          auto padrowEntry = pad->GetPadRowNumber(projZEntry);
+          auto padrowExit = pad->GetPadRowNumber(trkWork->getX());
+          if (padrowEntry != padrowExit) {
+            trkWork->setIsPadrowCrossing(iLayer);
```
Should not this be be validated by the presence of a tracklet in the neighbouring padrow?

martenole replied:
Yes, when thinking about it this would make more sense. We would probably also want the information about pad row crossing only in case a tracklet is attached. So then you could move this check a bit down to when a tracklet is actually attached to the track. And in that case do in addition the check whether another tracklet is present on the neighboring pad row. Since the tracklets are not sorted within one chamber this probably requires a loop over all tracklet of the given chamber

tdietel replied:
I think the check for tracklets in neighboring padrows is complementary: it would detect if more information is available that should be merged into the already attached tracklet. A similar argument could actually be made for tracklets in the same padrow, especially at low pT and large inclination angle.
The flagging based on tracking only would indicate if some signal was deposited on a neighboring padrow, even it is below the threshold required to generate a tracklet. Even in such a case, the quality of at least one time slice would be compromised, but it would not be detectable by looking for neighboring tracklets.

martenole replied:
> A similar argument could actually be made for tracklets in the same padrow, especially at low pT and large inclination angle.

Are you sure about this? I think low pT tracks spanning over more than 4 pads would rather not be found at all. And for the ones we find (>~700MeV) we should not loose charge information in column direction.

The problem I see when relying purely on the information from the track parameters is that especially for TPC tracks the estimate for the z position might be quite unprecise while eta should be rather well described. This means you can shift the track to some extent along z. And when doing that you can of course always find some position where the track crossed pad rows, but also a position where it did not. Checking if there really were 2 tracklets might be a clearer indication. But that would require a good efficiency for the online tracklets of course. Probably it would be good to check in Run 3 data how often this is the case that for a track two neighboring tracklets are in its vicinity. Could be done both in column and in row direction

tdietel replied:
> Are you sure about this? I think low pT tracks spanning over more than 4 pads would rather not be found at all. And for the ones we find (>~700MeV) we should not loose charge information in column direction.

I don't have numbers for this. I remember that a long time ago, we looked into tracklet splitting, and found that a significant number of tracklets (I remember 10-20%) had duplicates, i.e. were reconstructed twice in neighboring channels. But I do not know how how big the differences for the tracklet charges were.

tdietel replied:
> The problem I see when relying purely on the information from the track parameters is that especially for TPC tracks the estimate for the z position might be quite unprecise while eta should be rather well described. This means you can shift the track to some extent along z. And when doing that you can of course always find some position where the track crossed pad rows, but also a position where it did not.

That's a good point. Do you think this would also be a problem for ITS-TPC-TRD tracks? I guess in this case, the timing information should be good enough to fix z?

And to which degree does the TRD fix the z-position of the TPC-TRD tracks? I think we only use 

> Checking if there really were 2 tracklets might be a clearer indication. But that would require a good efficiency for the online tracklets of course. Probably it would be good to check in Run 3 data how often this is the case that for a track two neighboring tracklets are in its vicinity. Could be done both in column and in row direction

It seems that both checks test slightly different things, and both could be useful in different scenarios, e.g. ITS-TPC-TRD tracks without padrow crossings for gain calibration, where you prefer high-quality tracks, and a check for neighboring tracklets for PID, where you are more interested in efficiency.

I would therefore implement both.

f3sch replied:
> ITS-TPC-TRD tracks without padrow crossings for gain calibration, where you prefer high-quality tracks

Do you(@tdietel) want to generally throw away tracks if they have a padrow crossing for gain calibration or would you like to have this layerwise?

martenole commented on GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:
```diff
@@ -657,6 +657,22 @@ GPUd() bool GPUTRDTracker_t<TRDTRK, PROP>::FollowProlongation(PROP* prop, TRDTRK
+          // attached to the Track. It is estimated that the probability for
+          // this is low.
+          auto padrowEntry = pad->GetPadRowNumber(projZEntry);
+          auto padrowExit = pad->GetPadRowNumber(trkWork->getX());
```
```suggestion
          auto padrowExit = pad->GetPadRowNumber(trkWork->getZ());
```

f3sch replied:
Missed that thanks.

martenole commented on GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:
```diff
@@ -657,6 +657,22 @@ GPUd() bool GPUTRDTracker_t<TRDTRK, PROP>::FollowProlongation(PROP* prop, TRDTRK
+        if (iDet == 0) {
+          float projZEntry, projYEntry;
+          // Get Z for Entry of Track
+          prop->getPropagatedYZ(trkWork->getX() - 3.f, projYEntry, projZEntry);
```
The drift length of 3cm you should be able to find in the geometry

f3sch replied:
Ok.

martenole commented on GPU/GPUTracking/DataTypes/GPUTRDTrack.h:
```diff
@@ -86,6 +86,7 @@ class GPUTRDTrack_t : public T
   GPUd() short getCollisionId() const { return mCollisionId; }
   GPUd() int getNtracklets() const;
   GPUd() float getChi2() const { return mChi2; }
+  GPUd() unsigned char getIsPadrowCrossing() const { return mIsPadrowCrossing; }
```
Maybe you could also add a getter `bool getIsPadrowCrossing(int layer)` for convenience?

martenole commented on GPU/GPUTracking/DataTypes/GPUTRDTrack.cxx:
```diff
@@ -31,6 +31,7 @@ GPUd() void GPUTRDTrack_t<T>::initialize()
 {
   // set all members to their default values (needed since in-class initialization not possible with AliRoot)
   mChi2 = 0.f;
+  mIsPadrowCrossing = 0;
```
There are more functions in `GPUTRDTrack` where the `mPadRowCrossing` variable should be initialized. It should in addition be added to `GPU/GPUTracking/TRDTracking/GPUTRDTrackData.h`

f3sch replied:
Think I got them all...

martenole commented on Detectors/TRD/qc/include/TRDQC/Tracking.h:
```diff
@@ -42,15 +42,16 @@ namespace trd
+  float pt;                       ///< the transverse momentum of the track at the point of the innermost ITS cluster (ITS-TPC-TRD) or at the inner TPC radius (TPC-TRD)
+  float ptSigma2;                 ///< Sigma2 of pt
+  float dEdxTotTPC;               ///< raw total dEdx information for seeding track in TPC
+  unsigned char isPadrowCrossing; ///< bits 0 to 5 indicate if track crossed a padrow in that layer
```
In here you could also use a bitset, since this data type will not be used on the GPU

alibuild commented:
Error while checking build/O2/fullCI for f0132ce041d2a8a643669895e2994188c1d2693c at 2022-11-08 11:04:
```
## sw/BUILD/O2-latest/log
lld: error: undefined hidden symbol: o2::trd::PadPlane::getPadRowNumber(double) const
clang-14: error: amdgcn-link command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10229/f0132ce041d2a8a643669895e2994188c1d2693c/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for 26370095242e04fc084f360bcd985f4a2f70d6f1 at 2022-11-08 18:41:
```
## sw/BUILD/O2-latest/log
lld: error: undefined hidden symbol: o2::trd::PadPlane::getPadRowNumber(double) const
clang-14: error: amdgcn-link command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10229/26370095242e04fc084f360bcd985f4a2f70d6f1/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for 01d9c4c82e1600bf39234031405fd2b41fa21b42 at 2022-11-09 19:53:
```
## sw/BUILD/O2-latest/log
lld: error: undefined hidden symbol: o2::trd::PadPlane::getPadRowNumber(double) const
clang-14: error: amdgcn-link command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10229/01d9c4c82e1600bf39234031405fd2b41fa21b42/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for 132e18eec2232ae5407fdc41f28e2eedfe70c68d at 2022-11-11 12:49:
```
## sw/BUILD/O2-latest/log
lld: error: undefined hidden symbol: o2::trd::PadPlane::getPadRowNumber(double) const
clang-14: error: amdgcn-link command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10229/132e18eec2232ae5407fdc41f28e2eedfe70c68d/build_O2_fullCI/pretty.html).

martenole commented on Detectors/TRD/qc/include/TRDQC/Tracking.h:
```diff
@@ -42,15 +43,16 @@ namespace trd
+  float pt;                        ///< the transverse momentum of the track at the point of the innermost ITS cluster (ITS-TPC-TRD) or at the inner TPC radius (TPC-TRD)
+  float ptSigma2;                  ///< Sigma2 of pt
+  float dEdxTotTPC;                ///< raw total dEdx information for seeding track in TPC
+  std::bitset<6> isPadrowCrossing; ///< indicate if track crossed a padrow in that layer
```
Maybe you could directly add the global track ID as well?

f3sch replied:
I am still figuring the TOF id out. Let us not intermix features/request across PRs and first get this merged.

martenole replied:
Well it would be just adding one more integer to the QC struct but as you wish ;)

martenole commented on GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:
```diff
@@ -821,6 +821,33 @@ GPUd() bool GPUTRDTracker_t<TRDTRK, PROP>::FollowProlongation(PROP* prop, TRDTRK
+        if (mHypothesis[iUpdate + hypothesisIdxOffset].mTrackletId == trkltIdx) {
+          continue;
+        }
+        if (GPUCommonMath::Abs(tracklets[mHypothesis[iUpdate + hypothesisIdxOffset].mTrackletId].GetZbin() - tracklets[trkltIdx].GetZbin())) {
```
Should you not also check the distance along pad column direction? And the check the check in row direction also does not look correct. You should check that the distance in row direction is exactly 1 pad

f3sch replied:
Good catch!

f3sch committed:
TRD: Tracker: Indicate that there was neighboring tracklet

This helps later to identify whether z-row merging should be done to
recover the full charge information.

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

f3sch commented:
I could reproduce your numbers for the crossings and neighboring tracklets, after running the tracking:
![image](https://user-images.githubusercontent.com/42321018/202024879-316d4309-8d26-4dcd-918d-dbbf7ee3036b.png)
So the code should now be ready to be merged.

martenole commented:
And what is the fraction now with the code as is in this PR? Should be less than 40% now, right?

f3sch commented:
The numbers above are with the current PR, my original numbers were a combination of two things 1. without checking the distances in padrow and 2. me not recompiling O2PDPSuite... hence not linking to the new O2 compilation (I think...)

martenole commented:
so now its 11% with the combined condition `padRowEntry != padRowExit || neighbouringTrackletAvailable`?

martenole commented on GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:
```diff
@@ -821,6 +821,35 @@ GPUd() bool GPUTRDTracker_t<TRDTRK, PROP>::FollowProlongation(PROP* prop, TRDTRK
+      const auto padrowExit = pad->GetPadRowNumber(trkWork->getZ());
+      if (padrowEntry != padrowExit) {
+        trkWork->setIsPadrowCrossing(iLayer);
+        continue; // we can only mark padrow crossings once, hence the next loop is pointless
```
I guess with the `hasNeighbor` check one should still do the check below, right?

f3sch replied:
yes, will fix and report back with new numbers.

f3sch committed:
TRD: Tracker: Indicate if a padrow crossing was found

This removes a previous bug, where when a padrow crossing was found the
check for neighboring tracklets was skipped. Additionally, the 7th bit
is used to indicate whether or not a padrow in a track was found.

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

f3sch commented:
Finally got the numbers done, here in matrix form:
|          | Crossings (%) | Neighbors (%) |
|----------|--------------|--------------|
| Both     | 11.9         | 9.5          | 
| Crossings | 11.4         | 0.0          |
| Neighbors | 0.0          | 8.8          |

-> there is of course overlap show in Both
from 100 TF  with 522829 tracklets
